File: algoexpert/easy/python/Branch_sum.py
class Node:
	def __init__(self, value):
		self.val = value
		self.left = None
		self.right = None


# Probably we gonna do it using recursive
# O(n) time | O(log(n)) space complexity

def BranchSum(root):
	sums = []
	_calculateBranchSum(root, 0, sums)
	return sums

def _calculateBranchSum(node, runningSum, sums):
	if node is None:
		return
	newRunningSum = runningSum + node.val
	if node.left is None and node.rigth is None:
		sums.append(newRunningSum)
		return
	return _calculateBranchSum(node.left, newRunningSum, sums)
	return _calculateBranchSum(node.right, newRunningSum, sums)

File: algoexpert/easy/javascript/Branch_sum.js
class Node {
	constructor(value) {
		this.value = value;
		this.left = null;
		this.right = null;
	}
}

const branchSum = (root) => {
	let sums = [];
	_calculateBranchSum(root, 0, sums);
	return sums;
}

const _calculateBranchSum = (node, runningSum, sums) => {
	if (node === null) return;
	let newRunningSum = runningSum + node.value;
	if (node.left === null && node.right === null) {
		sums.push(newRunningSum);
		return;
	}
	return _calculateBranchSum(node.left, newRunningSum, sums),
		_calculateBranchSum(node.right, newRunningSum, sums);
}